From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Sun, 16 Feb 2025 15:25:45 +0800
Subject: [PATCH 01/20] dt-bindings: Add Apple/Dialog PMICs and subdevice
 bindings

Add bindings for Dialog PMICs and subdevices present in Apple products with
A7 - A11, T2 SoCs.

The PMIC device will be exposed via a simple-mfd driver either on
I2C or SPMI, so generic compatibles "apple,i2c-pmic" and
"apple,spmi-pmic" are used as such bindings could describe arbitrary
hardware attached via either I2C or SPMI

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 .../leds/backlight/apple,pmic-bl.yaml         |  41 +++++
 .../devicetree/bindings/mfd/apple,pmic.yaml   | 152 ++++++++++++++++++
 .../bindings/nvmem/apple,pmic-nvmem.yaml      |  37 +++++
 .../bindings/rtc/apple,pmic-rtc.yaml          |  65 ++++++++
 4 files changed, 295 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/leds/backlight/apple,pmic-bl.yaml
 create mode 100644 Documentation/devicetree/bindings/mfd/apple,pmic.yaml
 create mode 100644 Documentation/devicetree/bindings/nvmem/apple,pmic-nvmem.yaml
 create mode 100644 Documentation/devicetree/bindings/rtc/apple,pmic-rtc.yaml

diff --git a/Documentation/devicetree/bindings/leds/backlight/apple,pmic-bl.yaml b/Documentation/devicetree/bindings/leds/backlight/apple,pmic-bl.yaml
new file mode 100644
index 000000000..34cc7d545
--- /dev/null
+++ b/Documentation/devicetree/bindings/leds/backlight/apple,pmic-bl.yaml
@@ -0,0 +1,41 @@
+# SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/leds/backlight/apple,pmic-bl.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Apple PMIC backlight
+
+maintainers:
+  - Nick Chan <towinchenmi@gmail.com>
+
+description:
+  This module is part of Apple PMICs. For more details see
+  ../mfd/apple,pmic.yaml.
+
+  A backlight WLED interface in Apple PMICs. There are two variants
+  differing in how the backlight value is encoded in the registers.
+
+allOf:
+  - $ref: common.yaml#
+
+properties:
+  compatible:
+    oneOf:
+      - items:
+          - const: apple,arabela-pmic-bl
+          - const: apple,anya-pmic-bl
+      - items:
+          - const: apple,august-pmic-bl
+          - const: apple,aria-pmic-bl
+      - const: apple,anya-pmic-bl
+      - const: apple,aria-pmic-bl
+
+  reg:
+    maxItems: 1
+
+required:
+  - compatible
+  - reg
+
+additionalProperties: false
diff --git a/Documentation/devicetree/bindings/mfd/apple,pmic.yaml b/Documentation/devicetree/bindings/mfd/apple,pmic.yaml
new file mode 100644
index 000000000..8bf64a835
--- /dev/null
+++ b/Documentation/devicetree/bindings/mfd/apple,pmic.yaml
@@ -0,0 +1,152 @@
+---
+$id: http://devicetree.org/schemas/mfd/apple,pmic.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Apple PMICs
+
+maintainers:
+  - Nick Chan <towinchenmi@gmail.com>
+
+description: |
+  Apple SoCs came with PMICs designed by Dialog Semiconductor before
+  Apple switched to their own PMIC designs. However, the parts used
+  by Apple appears to be sigificantly different from public parts
+  offered by dialog.
+
+  On A7 - A10X SoCs, they are attached over I2C. On A11 and T2 the
+  PMIC is attached via SPMI.
+
+  This binding describes simple MFD - I2C/SPMI device with 16-bit
+  register address and 8-bit values that allows subdevices like RTC,
+  power-off control, backlight and nvmem to be described.
+
+  See also bindings/nvmem/apple,spmi-nvmem.yaml for a simpler
+  binding with only nvmem functions exposed used for newer PMICs.
+
+properties:
+  compatible:
+    oneOf:
+      - items:
+          - enum:
+              - apple,adelyn-pmic
+              - apple,adi-pmic
+              - apple,amber-pmic
+              - apple,angel-pmic
+              - apple,antigua-pmic
+              - apple,anya-pmic
+              - apple,arabela-pmic
+              - apple,aria-pmic
+              - apple,august-pmic
+              - apple,autumn-pmic
+          - const: apple,i2c-pmic
+      - items:
+          - enum:
+              - apple,agnes-pmic
+              - apple,calpe-pmic
+          - const: apple,spmi-pmic
+
+  reg:
+    maxItems: 1
+
+  "#address-cells":
+    const: 1
+
+  "#size-cells":
+    const: 1
+
+patternProperties:
+  "^backlight(@[0-9a-f]+)?$":
+    $ref: /schemas/leds/backlight/apple,pmic-bl.yaml
+
+  "^rtc(@[0-9a-f]+)?$":
+    $ref: /schemas/rtc/apple,pmic-rtc.yaml
+
+  "^nvmem(@[0-9a-f]+)?$":
+    $ref: /schemas/nvmem/apple,pmic-nvmem.yaml
+
+required:
+  - compatible
+  - reg
+  - "#address-cells"
+  - "#size-cells"
+
+additionalProperties: false
+
+examples:
+  - |
+    i2c {
+        #address-cells = <1>;
+        #size-cells = <0>;
+
+        pmic@3c {
+            compatible = "apple,arabela-pmic", "apple,i2c-pmic";
+            reg = <0x3c>;
+            #address-cells = <1>;
+            #size-cells = <1>;
+
+            rtc@500 {
+                compatible = "apple,arabela-pmic-rtc", "apple,amber-pmic-rtc";
+                reg = <0x500 0x10>;
+                nvmem-cells = <&rtc_offset_0>;
+                nvmem-cell-names = "rtc_offset";
+            };
+
+            backlight@600 {
+                compatible = "apple,arabela-pmic-bl", "apple,anya-pmic-bl";
+                reg = <0x600 0x2>;
+            };
+
+            nvmem@4000 {
+                compatible = "apple,pmic-nvmem";
+                reg = <0x4000 0x28>;
+
+                nvmem-layout {
+                    compatible = "fixed-layout";
+                    #address-cells = <1>;
+                    #size-cells = <1>;
+
+                    rtc_offset_0: rtc-offset@4 {
+                        reg = <0x4 0x4>;
+                    };
+                };
+            };
+        };
+    };
+
+  - |
+
+    #include <dt-bindings/spmi/spmi.h>
+
+    spmi {
+        #address-cells = <2>;
+        #size-cells = <0>;
+
+        pmic@f {
+            compatible = "apple,agnes-pmic", "apple,spmi-pmic";
+            reg = <0xf SPMI_USID>;
+            #address-cells = <1>;
+            #size-cells = <1>;
+
+            rtc@1400 {
+                compatible = "apple,agnes-pmic-rtc", "apple,antigua-pmic-rtc";
+                reg = <0x1400 0x10>;
+                nvmem-cells = <&rtc_offset_1>;
+                nvmem-cell-names = "rtc_offset";
+            };
+
+            nvmem@6000 {
+                compatible = "apple,pmic-nvmem";
+                reg = <0x6000 0x300>;
+
+                nvmem-layout {
+                    compatible = "fixed-layout";
+                    #address-cells = <1>;
+                    #size-cells = <1>;
+
+                    rtc_offset_1: rtc-offset@4 {
+                        reg = <0x4 0x4>;
+                    };
+                };
+            };
+        };
+    };
diff --git a/Documentation/devicetree/bindings/nvmem/apple,pmic-nvmem.yaml b/Documentation/devicetree/bindings/nvmem/apple,pmic-nvmem.yaml
new file mode 100644
index 000000000..b184c10ed
--- /dev/null
+++ b/Documentation/devicetree/bindings/nvmem/apple,pmic-nvmem.yaml
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: GPL-2.0-only OR BSD-2-Clause
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/nvmem/apple,pmic-nvmem.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Apple PMIC NVMEM
+
+description: |
+  This module is part of Apple PMICs. For more details see
+  ../mfd/apple,pmic.yaml.
+
+  NVMEM cells found in Apple PMICs are used to store information
+  including RTC offset and boot failure count.
+
+  See also bindings/nvmem/apple,spmi-nvmem.yaml which describes
+  PMICs with only NVMEM functions exposed.
+
+maintainers:
+  - Nick Chan <towinchenmi@gmail.com>
+
+allOf:
+  - $ref: nvmem.yaml#
+
+properties:
+  compatible:
+    const: apple,pmic-nvmem
+
+  reg:
+    maxItems: 1
+
+required:
+  - compatible
+  - reg
+
+unevaluatedProperties: false
+...
diff --git a/Documentation/devicetree/bindings/rtc/apple,pmic-rtc.yaml b/Documentation/devicetree/bindings/rtc/apple,pmic-rtc.yaml
new file mode 100644
index 000000000..de0635797
--- /dev/null
+++ b/Documentation/devicetree/bindings/rtc/apple,pmic-rtc.yaml
@@ -0,0 +1,65 @@
+# SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/rtc/apple,pmic-rtc.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Apple PMIC RTC device
+
+maintainers:
+  - Nick Chan <towinchenmi@gmail.com>
+
+description:
+  This module is part of Apple PMICs. For more details
+  see ../mfd/apple,pmic.yaml.
+
+  An RTC device that is part of the PMIC. Additionally an
+  offset stored in an nvmem cell is required to compute the
+  current date and time.
+
+properties:
+  compatible:
+    oneOf:
+      - items:
+          - enum:
+              - apple,adi-pmic-rtc
+              - apple,anya-pmic-rtc
+              - apple,arabela-pmic-rtc
+          - const: apple,amber-pmic-rtc
+      - items:
+          - enum:
+              - apple,aria-pmic-rtc
+              - apple,adelyn-pmic-rtc
+              - apple,angel-pmic-rtc
+              - apple,august-pmic-rtc
+              - apple,agnes-pmic-rtc
+              - apple,calpe-pmic-rtc # Counter also exposed via SMC
+          - const: apple,antigua-pmic-rtc
+      - const: apple,amber-pmic-rtc
+      - const: apple,antigua-pmic-rtc
+
+  reg:
+    maxItems: 1
+
+  interrupts:
+    maxItems: 1
+
+  nvmem-cells:
+    items:
+      - description:
+          Upper 32 Bit of RTC offset in seconds.
+          # TODO: another cell for ticks specified in 32768 (2^15) Hz
+          # clock ticks
+
+  nvmem-cell-names:
+    items:
+      - const: rtc_offset
+
+  wakeup-source: true
+
+required:
+  - compatible
+  - reg
+
+additionalProperties: false
+...
